plasmapy/utils/data/downloader.py

```diff
@@ -0,0 +1,76 @@
+"""
+Contains functionality for downloading files from a URL. Intended for
+downloading files from the PlasmaPy data repository.
+
+"""
+
+import os
+import requests
+
+from urllib.parse import urljoin
+
+__all__ = ["get_file"]
+
+# Note: GitHub links have a problem where the Content-Encoding is set to
+# 'gzip' but the file is not actually compressed. This header is just ignored
+# by the get_file function.
+_BASE_URL = "https://github.com/PlasmaPy/PlasmaPy-data/raw/main/data/"
+
+# TODO: use a config file variable to allow users to set a location for this folder?
+_DOWNLOADS_PATH = os.path.join(os.path.dirname(__file__), "downloads")
+
+
+def get_file(basename, base_url=_BASE_URL, directory=None):
+    r"""
+    Downloads a file from a URL (if the file does not already exist) and
+    returns the full local path to the file.
+
+    Parameters
+    ----------
+    basename : str
+        Name of the file to be downloaded (extension included).
+
+    base_url : str, optional
+        The base URL of the file to be downloaded. Defaults to the main
+        directory of `_PlasmaPy's data repository`.
+
+    directory : str, optional
+        The full path to the desired download location. Defaults to the
+        default PlasmaPy data download directory:
+        plasmapy/utils/data/downloads/
+
+    Returns
+    -------
+    path : str
+        The full local path to the downloaded file.
+
+    """
+
+    if "." not in str(basename):
+        raise ValueError(f"'filename' ({basename}) must include an extension.")
+
+    if directory is None:  # coverage: ignore
+        directory = _DOWNLOADS_PATH
+
+    path = os.path.join(directory, basename)
+
+    # If file doesn't exist locally, download it
```

Comment:
One limitation of this that I'm sure has already been solved elsewhere is that if the upstream file changes, but does not change the filename, this will not update the local version. Could be good to note that in the docstring!

One way of solving that in the future would be to note the SHA1 checksum in the filename, e.g. store "NIST.dat" as "NIST-1234567890.dat" online, and if the file exists... Yeah, this is getting a little too complex for this train ride; it'd be good to investigate how sunpy or astropy or sklearn solve this. Future work, though.

Reply:
I've been playing around with this on [another branch](https://github.com/PlasmaPy/PlasmaPy-data/tree/data_with_hashes), but I think we should leave it for a future PR. For now this module is exclusively for notebooks + testing, so we can just manage it manually for now.